fdbserver/DataDistribution.actor.cpp

```diff
@@ -1512,219 +1631,690 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
+ACTOR Future<Void> loadAndDispatchAudit(Reference<DataDistributor> self,
+                                        std::shared_ptr<DDAudit> audit,
+                                        KeyRange range) {
+	TraceEvent(SevInfo, "DDLoadAndDispatchAudit", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("AuditType", audit->coreState.getType());
+	if (audit->coreState.getType() == AuditType::ValidateStorageServerShard) {
+		audit->actors.add(auditInputRangeOnAllStorageServers(self, audit, allKeys));
+	} else if (audit->coreState.getType() == AuditType::ValidateLocationMetadata) {
+		audit->actors.add(auditMakeProgressOnRange(self, audit, allKeys));
+		// audit->actors.add(runAuditJobOnOneRandomServer(self, audit, allKeys));
+	} else if (audit->coreState.getType() == AuditType::ValidateHA ||
+	           audit->coreState.getType() == AuditType::ValidateReplica) {
+		audit->actors.add(auditMakeProgressOnRange(self, audit, range));
+	} else {
+		UNREACHABLE();
+	}
+	wait(delay(0.1));
+	return Void();
+}
+
+// Randomly pick a server to run an audit on the input range
+ACTOR Future<Void> runAuditJobOnOneRandomServer(Reference<DataDistributor> self,
+                                                std::shared_ptr<DDAudit> audit,
+                                                KeyRange range) {
+	ASSERT(audit->coreState.getType() == AuditType::ValidateLocationMetadata);
+	TraceEvent(SevInfo, "DDRunAuditJobBySingleServerBegin", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("AuditType", audit->coreState.getType());
+	try {
+		ServerWorkerInfos serverWorkers = wait(self->txnProcessor->getServerListAndProcessClasses());
+		int selected = deterministicRandom()->randomInt(0, serverWorkers.servers.size());
+		audit->actors.add(auditMakeProgressOnServer(
+		    self, audit, range, serverWorkers.servers[selected].first, /*makeProgressByServer=*/false));
+	} catch (Error& e) {
+		TraceEvent(SevWarn, "DDRunAuditJobBySingleServerError", self->ddId)
+		    .errorUnsuppressed(e)
+		    .detail("AuditID", audit->coreState.id)
+		    .detail("AuditType", audit->coreState.getType());
+		throw e;
+	}
+	TraceEvent(SevInfo, "DDRunAuditJobBySingleServerEnd", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("AuditType", audit->coreState.getType());
+
+	return Void();
+}
+
+// For each of storage servers, run an audit on the input range
+ACTOR Future<Void> auditInputRangeOnAllStorageServers(Reference<DataDistributor> self,
+                                                      std::shared_ptr<DDAudit> audit,
+                                                      KeyRange range) {
+	ASSERT(audit->coreState.getType() == AuditType::ValidateStorageServerShard);
+	TraceEvent(SevInfo, "DDAuditInputRangeOnAllStorageServersBegin", self->ddId)
+	    .detail("AuditID", audit->coreState.id)
+	    .detail("AuditType", audit->coreState.getType());
+	state StorageServerInterface targetServer;
```

Comment:
nit: we don't need a `state` variable for `targetServer`, it can be a local variable in the for loop